proxy/context.go

```diff
@@ -233,10 +237,104 @@ func (c *context) clone() *context {
 	return &cc
 }
 
+func (c *context) wasExecuted()bool {
+	return c.executionCounter != 0
+}
+
 func (c *context) setMetricsPrefix(prefix string) {
 	c.metrics.prefix = prefix + ".custom."
 }
 
+func (tt *teeTie) Read(b []byte) (int, error) {
+	n, err := tt.r.Read(b)
+
+	if err != nil && err != io.EOF {
+		tt.w.CloseWithError(err)
+		return n, err
+	}
+
+	if n > 0 {
+		if _, werr := tt.w.Write(b[:n]); werr != nil {
+			log.Error("tee: error while tee request", werr)
+		}
+	}
+
+	if err == io.EOF {
+		tt.w.Close()
+	}
+
+	return n, err
+}
+
+func (tt *teeTie) Close() error { return nil }
+
+// Returns the cloned request and the tee body to be used on the main request.
+func cloneRequest(u *url.URL, req *http.Request) (*http.Request, io.ReadCloser, error) {
+	h := make(http.Header)
+	for k, v := range req.Header {
+		h[k] = v
+	}
+
+	var teeBody io.ReadCloser
+	mainBody := req.Body
+
+	// see proxy.go:231
+	if req.ContentLength != 0 {
+		pr, pw := io.Pipe()
+		teeBody = pr
+		mainBody = &teeTie{mainBody, pw}
+	}
+
+	clone, err := http.NewRequest(req.Method, u.String(), teeBody)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	clone.Header = h
+	clone.ContentLength = req.ContentLength
+
+	return clone, mainBody, nil
+}
+
+func (c *context) Split() (filters.FilterContext, error) {
+	originalRequest := c.Request()
+	cc := c.clone()
+	c.stateBag = map[string]interface{}{}
+	cc.metrics = &filterMetrics{
+		prefix: cc.metrics.prefix,
+		impl:   cc.proxy.metrics,
+	}
+	u := new(url.URL)
+	*u = *originalRequest.URL
+	cr, body, err := cloneRequest(u, originalRequest)
+	if err != nil {
+		log.Errorf("context: failed to clone request: %v", err)
+		return nil, err
+	}
+	originalRequest.Body = body
+	cc.request = cr
+	return cc, nil
+}
+
+func (c *context) Loopback() {
+	err := c.proxy.do(c)
+	if c.response != nil && c.response.Body != nil{
+		if _, err := io.Copy(ioutil.Discard, c.response.Body); err != nil {
+			c.proxy.log.Errorf("context: error while discarding remainder response body: %v.", err)
+		}
+		err := c.response.Body.Close()
+		if err != nil {
+			c.proxy.log.Error("context: error during closing the response body", err)
+		}
+	}
+	if c.proxySpan != nil {
+		c.proxySpan.Finish()
+	}
+	if err != nil {
+		log.Error("context: failed to execute loopback request", err)
```

Comment:
we have some logging enhancements in the Proxy.errorResponse function. Some of it should also be applied here (flow ID. It's fine with me in a second pass, though.

Reply:
We can postpone this for the next iteration.